cmd/bench/main.go

```diff
@@ -0,0 +1,225 @@
+	} else {
+		// Run a latency test
+		var clients []bclient.LatencyClient
+		for i := 0; i < workersFlag; i++ {
+			op := makeECDSASignOp(params.ECDSASHA512Params, ski)
+			c, err := makeLatencyClientFromOp(cli, serverFlag, fmt.Sprint(portFlag), op)
+			if err != nil {
+				panic(err)
+			}
+			clients = append(clients, c)
+		}
+
+		buckets := bclient.RunLatencyClients(durFlag, minFlag, maxFlag, stepFlag, clients...)
+		bclient.PrintHistogram(buckets)
+	}
+}
+
+type bwClient struct {
+	pkt  []byte
+	conn *tls.Conn
+}
+
+func (b *bwClient) Dispatch() {
+	_, err := b.conn.Write(b.pkt)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (b *bwClient) Complete() {
+	readPacket(b.conn)
+}
+
+func readPacket(conn *tls.Conn) {
+	var buf [8]byte
+	_, err := io.ReadFull(conn, buf[:])
+	if err != nil {
+		panic(err)
+	}
+
+	var pkt gokeyless.Packet
+	err = pkt.UnmarshalBinary(buf[:])
+	if err != nil {
+		panic(err)
+	}
+
+	body := make([]byte, pkt.Length)
+	_, err = io.ReadFull(conn, body)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func makeBandwidthClientFromOp(cli *client.Client, server, port string, op *gokeyless.Operation) (bclient.BandwidthClient, error) {
+	conn := dial(cli, server, port)
+	pkt, err := gokeyless.NewPacket(op).MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+	return &bwClient{
+		pkt:  pkt,
+		conn: conn,
+	}, nil
+}
+
+func makeLatencyClientFromOp(cli *client.Client, server, port string, op *gokeyless.Operation) (bclient.LatencyClient, error) {
+	conn := dial(cli, server, port)
+	pkt, err := gokeyless.NewPacket(op).MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+	return bclient.FuncLatencyClient(func() time.Duration {
+		t0 := time.Now()
+		_, err := conn.Write(pkt)
+		if err != nil {
+			panic(err)
+		}
+		readPacket(conn)
+		t1 := time.Now()
+		if t1.Before(t0) {
+			// Time went backwards (probably because somebody reset the clock - Go
+			// doens't yet support monotonic clocks). This test is obviously invalid.
+			panic("time moved backwards; the results of this test are untrustworthy")
+		}
+
+		time.Sleep(pauseFlag)
+
+		return t1.Sub(t0)
+	}), nil
+}
+
+// server must be the TLS name of the server and a resolvable domain name.
+func dial(cli *client.Client, server, port string) *tls.Conn {
+	config := cli.Config.Clone()
+	config.ServerName = server
+	conn, err := tls.Dial("tcp", server+":"+port, config)
+	if err != nil {
+		panic(err)
+	}
+	return conn
+}
+
+func makeRSASignOp(params params.RSASignParams, SKI gokeyless.SKI) *gokeyless.Operation {
+	return &gokeyless.Operation{
+		Opcode:  params.Opcode,
+		Payload: randBytes(params.PayloadSize),
+		SKI:     SKI,
+	}
+}
+
+func makeECDSASignOp(params params.ECDSASignParams, SKI gokeyless.SKI) *gokeyless.Operation {
+	return &gokeyless.Operation{
+		Opcode:  params.Opcode,
+		Payload: randBytes(params.PayloadSize),
+		SKI:     SKI,
+	}
+}
+
+func randBytes(n int) []byte {
+	b := make([]byte, n)
+	_, err := io.ReadFull(rand.New(rand.NewSource(time.Now().UnixNano())), b)
```

Comment:
I have concerns though about that repeatedly reseeding a new rand stream is no better than using one rand stream. For efficiency, we can generate the source once each run and use it for the run time. What do think?